File: statistical_physics/ising_model_2d/julia/examples/ising_sim.jl
"""
2D Ising Model — Temperature Sweep (Julia)

Generates ising2d_sweep.csv with columns: T, E_mean, M_mean, Cv, chi

Usage:
  julia --project=statistical_physics/ising_model_2d/julia \\
        statistical_physics/ising_model_2d/julia/examples/ising_sim.jl
"""

push!(LOAD_PATH, joinpath(@__DIR__, "..", "src"))
using IsingModel2D

const N       = 32
const N_THERM = 5_000
const N_MEAS  = 10_000
const T_MIN   = 1.0
const T_MAX   = 4.0
const N_T     = 40

println("2D Ising Model sweep  N=$N  T=[$T_MIN, $T_MAX]  $N_T points")
println("Onsager T_c = $(round(T_CRITICAL, digits=4))")

temps = range(T_MIN, T_MAX, length=N_T)

out_file = joinpath(@__DIR__, "ising2d_sweep.csv")
open(out_file, "w") do io
    println(io, "T,E_mean,M_mean,Cv,chi")
    for (k, T) in enumerate(temps)
        m = IsingModel(N; seed=k)
        if T < T_CRITICAL
            fill!(m.lattice, Int8(1))   # start ordered below Tc
        end
        r = simulate(m, Float64(T); n_therm=N_THERM, n_measure=N_MEAS)
        println(io, "$(round(T, digits=4)),$(round(r.E_mean, digits=6))," *
                    "$(round(r.M_mean, digits=6)),$(round(r.Cv, digits=6))," *
                    "$(round(r.chi, digits=6))")
        if k % 10 == 0
            println("  $k/$N_T  T=$(round(T, digits=2))  |M|=$(round(r.M_mean, digits=3))  Cv=$(round(r.Cv, digits=3))")
        end
    end
end
println("\nSaved → $out_file")

File: statistical_physics/ising_model_2d/cpp/examples/ising_sim.cpp
/**
 * 2D Ising Model — Temperature Sweep
 *
 * Generates ising2d_sweep.csv with columns:
 *   T, E_mean, M_mean, Cv, chi
 *
 * Usage (from cpp/build/):
 *   cmake .. && cmake --build . && ./ising_sim
 */

#include "ising_model.hpp"

#include <cstdio>
#include <cmath>
#include <vector>

int main() {
    const int    N       = 32;
    const int    N_THERM = 5'000;
    const int    N_MEAS  = 10'000;
    const double T_MIN   = 1.0;
    const double T_MAX   = 4.0;
    const int    N_T     = 40;

    std::printf("2D Ising Model sweep  N=%d  T=[%.1f, %.1f]  %d points\n",
                N, T_MIN, T_MAX, N_T);
    std::printf("Onsager T_c = %.4f\n\n", T_CRITICAL);

    std::vector<double> temps(N_T);
    for (int k = 0; k < N_T; ++k)
        temps[k] = T_MIN + k * (T_MAX - T_MIN) / (N_T - 1);

    // Open CSV
    std::FILE* fp = std::fopen("ising2d_sweep.csv", "w");
    std::fprintf(fp, "T,E_mean,M_mean,Cv,chi\n");

    for (int k = 0; k < N_T; ++k) {
        double T = temps[k];
        IsingModel2D model(N, 1.0, static_cast<uint64_t>(k));
        // Start ordered below Tc
        if (T < T_CRITICAL)
            for (int i = 0; i < N; ++i)
                for (int j = 0; j < N; ++j)
                    model.lattice[i][j] = 1;

        SimResult r = model.simulate(T, N_THERM, N_MEAS);
        std::fprintf(fp, "%.4f,%.6f,%.6f,%.6f,%.6f\n",
                     T, r.E_mean, r.M_mean, r.Cv, r.chi);

        if ((k + 1) % 10 == 0)
            std::printf("  %d/%d  T=%.2f  |M|=%.3f  Cv=%.3f\n",
                        k + 1, N_T, T, r.M_mean, r.Cv);
    }
    std::fclose(fp);
    std::printf("\nSaved → ising2d_sweep.csv\n");
    return 0;
}
